Defer image decoding when scrolling fast

dnfield opened the pull request:
## Description

Continuation of https://github.com/flutter/flutter/pull/48536

Dependds on https://github.com/flutter/flutter/pull/49319

Basically an API cleanup of #48536


## Related Issues

#32143
fixes #44510
b/144232910

#48305
Fixes #48775

## Tests

I added the following tests:

Tests for new provider, new method on image cache (containsKey), image behavior in a scrolling list.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [ ] No, no existing tests failed, so this is *not* a breaking change.
- [x] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [x] I wrote a design doc: https://flutter.dev/go//deferred-image-decoding
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [x] I wrote a migration guide: https://github.com/flutter/website/pull/3622


<!-- Links -->
[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

dnfield committed:
recommendDeferredLoading

dnfield committed:
newline

dnfield committed:
another test

dnfield committed:
typo

dnfield committed:
Merge branch 'physics_fast' into image_provider_widget

dnfield committed:
start

dnfield commented on packages/flutter/lib/src/painting/image_cache.dart:
```diff
@@ -227,6 +227,11 @@ class ImageCache {
   }
 
+  /// Returns whether this [key] has been previously added by [putIfAbsent].
+  bool containsKey(Object key) {
```
This is a break.

dnfield commented on packages/flutter/lib/src/painting/image_provider.dart:
```diff
@@ -270,10 +270,19 @@ abstract class ImageProvider<T> {
-  /// method.
+  /// method. If they need to manage the actual resolution of the image, they
+  /// should override [resolveForStream] instead of this method.
+  @nonVirtual
```
This is a break.

Hixie replied:
why must it be nonVirtual? I mean, I get that it's useful for catching people who are using the API the old way, but is it strictly necessary? What if someone wants to take an entirely different approach?

dnfield replied:
Because if some other provider overrides it, its new logic will never get called by the new kind of provider I'm making that would wrap it.

I imagine other providers would have this problem. Plus the logic in it is very complicated.

Alternatively, we could update docs to say "if you override this, composed providers may not work and you should consider overriding useKey instead" - but I'm inclined to think people are more likely to see an analyzer warning than a doc comment.

dnfield committed:
Merge remote-tracking branch 'upstream/master' into image_provider_widget

dnfield committed:
fixes

Hixie commented on packages/flutter/lib/src/painting/image_provider.dart:
```diff
@@ -270,10 +270,19 @@ abstract class ImageProvider<T> {
+    return stream;
+  }
+
+  void _safeObtainKey(ImageConfiguration configuration, ImageStream stream) {
```
"_safeObtainKey" doesn't return a key, which is weird naming

Hixie commented on packages/flutter/lib/src/painting/image_provider.dart:
```diff
@@ -270,10 +270,19 @@ abstract class ImageProvider<T> {
   /// Subclasses should implement [obtainKey] and [load], which are used by this
-  /// method.
+  /// method. If they need to manage the actual resolution of the image, they
+  /// should override [resolveForStream] instead of this method.
```
docs are out of date

Hixie commented on packages/flutter/lib/src/painting/image_provider.dart:
```diff
@@ -270,10 +270,19 @@ abstract class ImageProvider<T> {
   /// This is the public entry-point of the [ImageProvider] class hierarchy.
```
please add a section to the class docs (`## foo`) that discusses the lifecycle of this class and the order in which the various methods are called, by whom, when, and so on. The class is complicated enough now that an overview is necessary.

goderbauer replied:
+1

Hixie commented on packages/flutter/lib/src/painting/image_provider.dart:
```diff
@@ -322,17 +331,36 @@ abstract class ImageProvider<T> {
+  ///
+  /// Subclasses should avoid calling [obtainKey] directly and instead override
+  /// this method, which makes sure appropriate error handling is in place to
+  /// notify callers and the framework.
```
this paragraph is confusing given the first paragraph

Hixie commented on packages/flutter/lib/src/painting/image_provider.dart:
```diff
@@ -322,17 +331,36 @@ abstract class ImageProvider<T> {
+  /// this method, which makes sure appropriate error handling is in place to
+  /// notify callers and the framework.
+  ///
+  /// It is safe to call [handleError] multiple times if multiple errors occur.
```
Does this mean "It is safe for the implementation of this method to call..." ?

Hixie commented on packages/flutter/lib/src/painting/image_provider.dart:
```diff
@@ -322,17 +331,36 @@ abstract class ImageProvider<T> {
+  ///
+  /// The default implementation uses the key to interact with the [ImageCache],
+  /// calling [ImageCache.putIfAbsent] and notifying listeners of the [stream].
+  /// Implementers that do not call super should
```
incomplete

Hixie commented on packages/flutter/lib/src/widgets/scroll_aware_image_provider.dart:
```diff
@@ -0,0 +1,173 @@
+  /// The [State] must not be null, and [State.mounted] must be true.
+  ScrollAwareContextProvider(this._state)
+      : assert(_state != null),
+        assert(_state.mounted, 'A ScrollAwareContextProvider was given a BuildContext for an Element that was never mounted.');
```
was never -> is not

Hixie commented on packages/flutter/lib/src/widgets/scroll_aware_image_provider.dart:
```diff
@@ -0,0 +1,173 @@
+    return true;
+  }
+
+  /// Do not touch this.  Use [_context].
```
Not clear who this doc line is for. It's a private, so presumably it's for us. Bet we touch it 7 lines lower.

dnfield replied:
It's meant to say don't touch it from the provider.  But it's probably not necessary, I'll just remove it.

Hixie commented on packages/flutter/lib/src/widgets/scroll_aware_image_provider.dart:
```diff
@@ -0,0 +1,173 @@
+    assert(key != null);
+    assert(handleError != null);
+
+    void deferredResolve([bool firstCall = false]) {
```
do you need this to be a nested function? looks like you could get the same results with a regular method, which would be less confusing to follow

Hixie commented on packages/flutter/lib/src/widgets/scroll_aware_image_provider.dart:
```diff
@@ -0,0 +1,173 @@
+  /// If [dispose] has been called, will return null.
+  ///
+  /// Otherwise, asserts the [_state] is still mounted and returns its context.
+  BuildContext get _context {
```
making this private means that nobody outside this file can use this class. if this class is valuable enough that we're making it public, let's make it reusable as well.

dnfield replied:
This is on purpose - if we make it reusable, it should probably go somewhere else and have a different name.  

The class has to be public because it's a public parameter to the constructor of ScrollAwareImageProvider.

goderbauer commented on packages/flutter/lib/src/painting/image_provider.dart:
```diff
@@ -322,17 +331,36 @@ abstract class ImageProvider<T> {
-    return stream;
+  }
+
+  /// Called when [resolve] has finished calling [obtainKey].
```
From the docs and method name its not clear what useKey is supposed to do if I were to override it.

goderbauer commented on packages/flutter/lib/src/widgets/scroll_aware_image_provider.dart:
```diff
@@ -0,0 +1,173 @@
+///
+///   1. If the returned [ImageStream] has been completed, end. This can happen
+///      if the caller sets the completer on the stream.
+///   1. If the [ImageCache] has a completer for the key for this image, ask the
```
nit: actually enumerate this 1, 2,3 etc. to make this slightly more pleasant to read in the actual source?

goderbauer commented on packages/flutter/lib/src/widgets/scroll_aware_image_provider.dart:
```diff
@@ -0,0 +1,173 @@
+/// [ImageCache], and does not attempt to re-acquire or decode images in the
+/// cache.
+///
+/// Calling [resolve] on this provider will cause it to obtain the image key
```
Document somewhere what it means in practical terms for the image when the enumeration below says that resolve just "ends".

goderbauer commented on packages/flutter/lib/src/widgets/scroll_aware_image_provider.dart:
```diff
@@ -0,0 +1,173 @@
+  /// The context that may or may not be enclosed by a scrollable.
+  ///
+  /// Once [ScrollAwareContextProvider.dispose] is called on this context,
+  /// the provider will
```
the suspense ... what will it?

goderbauer commented on packages/flutter/lib/src/widgets/scroll_aware_image_provider.dart:
```diff
@@ -0,0 +1,173 @@
+  final ImageProvider<T> imageProvider;
+
+  @override
+  void useKey(
```
I don't have any great ideas, but this method name is not very descriptive...

dnfield replied:
going with `resolveStreamForKey`.

goderbauer commented on packages/flutter/lib/src/widgets/scroll_aware_image_provider.dart:
```diff
@@ -0,0 +1,173 @@
+      }
+      // Something still wants this image, but check if the context is scrolling
+      // too fast before scheduling work that might never show on screen.
+      // Try to get to end of the frame callbacks of the next frame, and then
```
Why try to? What could stop you? (Other then the engine not requesting any more frames?

dnfield replied:
Someone else could call `scheduleFrameCallback` after us and then schedule a microtask, or the engine could stop requesting frames.

dnfield committed:
docs, refactor, review

dnfield commented:
I've updated the docs and moved the build context enclosing class to a new class called `DisposableBuildContext`.

Still waiting for the dependent PR to land so I can rebase it into this one and then mark this ready for review.

dnfield committed:
merge

dnfield commented:
Ready for review now!

dnfield committed:
fix bad import

dnfield committed:
bad merge

goderbauer commented on packages/flutter/lib/src/painting/image_provider.dart:
```diff
@@ -182,6 +182,42 @@ typedef DecoderCallback = Future<ui.Codec> Function(Uint8List bytes, {int cacheW
 ///
 /// The following image formats are supported: {@macro flutter.dart:ui.imageFormats}
 ///
+/// ## Image resolving
```
nit: Maybe just make this heading "Lifecycle"? Or "Lifecycle of resolving an image" or some such?

goderbauer commented on packages/flutter/lib/src/painting/image_provider.dart:
```diff
@@ -182,6 +182,42 @@ typedef DecoderCallback = Future<ui.Codec> Function(Uint8List bytes, {int cacheW
+///      This stream will be used to communicate back to the caller when the
+///      image is decoded and ready to display, or when an error occurs.
+///   2. Obtain the key for the image using [obtainKey].
+///      Since this method is asynchronous, but also can return synchronously,
```
This sentence is strange. I assume you mean it's an asychnronous method, but it may throw exceptions synchronously? There is no way it can actually return anything useful synchronous?

dnfield replied:
We use `SynchronousFuture`s for these, which look async but are actually return synchronously.

dnfield replied:
Reworded a bit.

goderbauer commented on packages/flutter/lib/src/painting/image_provider.dart:
```diff
@@ -270,10 +306,31 @@ abstract class ImageProvider<T> {
-  /// method.
+  /// method. If they need to change the implementation of [ImageStream] used,
+  /// they should override [createStream]. If they need to manage the actual
+  /// resolution of the image, they should override [resolveStreamForKey].
```
Maybe add a link to the lifecycle explanation of this method in the class comment?

dnfield replied:
Done - I think.

goderbauer commented on packages/flutter/lib/src/painting/image_provider.dart:
```diff
@@ -322,17 +379,42 @@ abstract class ImageProvider<T> {
-    return stream;
+  }
+
+  /// Called when [resolve] has finished calling [obtainKey].
```
Maybe: Called by [resolve] with the key returned by [obtainKey].

goderbauer commented on packages/flutter/lib/src/painting/image_provider.dart:
```diff
@@ -270,10 +306,31 @@ abstract class ImageProvider<T> {
+  /// Called by [resolve] to create the [ImageStream] it returns.
+  ///
+  /// Subclasses should override this instead of [resolve] if they need to
+  /// return some subclass of [ImageStream].
```
Maybe link to resolveStreamForKey from here as well since some of the work performed on Stream can/should happen there instead of here.

goderbauer commented on packages/flutter/lib/src/widgets/disposable_build_context.dart:
```diff
@@ -0,0 +1,73 @@
+
+  /// Marks the [BuildContext] as disposed.
+  ///
+  /// Creators of this object should call [dispose] when their [Element] is
```
should -> must?

goderbauer commented on packages/flutter/lib/src/widgets/scroll_aware_image_provider.dart:
```diff
@@ -0,0 +1,109 @@
+  }) : assert(context != null),
+       assert(imageProvider != null);
+
+  /// The context that may or may not be enclosed by a scrollable.
```
nit: by a [Scrollable]?

goderbauer commented on packages/flutter/test/widgets/scroll_aware_image_provider_test.dart:
```diff
@@ -0,0 +1,335 @@
+    return Scrollable.of(find.byType(TestWidget).evaluate().first).position;
+  }
+
+  testWidgets('DisposableBuildContext asserts on disposed state', (WidgetTester tester) async {
```
Maybe move the DisposableBuildContext tests into their own file since the impl has its own home as well.

goderbauer commented on packages/flutter/test/widgets/scroll_aware_image_provider_test.dart:
```diff
@@ -0,0 +1,335 @@
+    expect(state.mounted, true);
+
+    final DisposableBuildContext context = DisposableBuildContext(state);
+    expect(context.debugValidate(), true);
```
Instead of marking debugValidate visible for tests, could we just test the same by actually accessing the context property and checking that we do or don't get an assert?

(That would also prevent us from ever removing that assert from the context getter)

goderbauer commented on packages/flutter/test/widgets/scroll_aware_image_provider_test.dart:
```diff
@@ -0,0 +1,335 @@
+    velocities.add(velocity);
+    return super.recommendDeferredLoading(velocity, metrics, context);
+  }
+}
```
nit: add a newline

goderbauer commented on packages/flutter/lib/src/widgets/disposable_build_context.dart:
```diff
@@ -0,0 +1,73 @@
+  /// Always returns true, but will assert if [dispose] has not been called
+  /// but the state this is tracking is unmounted.
+  @visibleForTesting
+  bool debugValidate() {
```
as pointed out in a comment on the test: Can we keep this private and instead test if accessing `context` throws?

goderbauer reviewed (approved):
LGTM after comments are resolved.

dnfield committed:
Merge remote-tracking branch 'upstream/master' into image_provider_widget

dnfield committed:
review

dnfield reviewed (commented):
Thanks for the review!

We should not land this until cl/289271973 is landed internally so it's a soft break.  I'll work on a migration guide.

dnfield commented:
Internal CL has landed.

Gperez88 commented:
when they move this to stable branch?
